### ISSUE TITLE
service fails if triggered too quickly

### ISSUE THREAD

**Issue by cgwalters** [coreos/console-login-helper-messages#87]
I spun up an OCP 4.6 cluster in libvirt and noticed that the service had failed; I think the problem is our `.path` unit is triggering too often.

Shades of https://github.com/coreos/bootupd/blob/3e62c9db9dc06bfb0acacdf1e862b28d84d7d8fa/systemd/bootupd.service#L7

```
sh-4.4# systemctl --failed
  UNIT                                           LOAD   ACTIVE SUB    DESCRIPTION                                         
● console-login-helper-messages-issuegen.service loaded failed failed Generate console-login-helper-messages issue snippet
...
sh-4.4# systemctl status console-login-helper-messages-issuegen.service
● console-login-helper-messages-issuegen.service - Generate console-login-helper-messages issue snippet
   Loaded: loaded (/usr/lib/systemd/system/console-login-helper-messages-issuegen.service; enabled; vendor preset: enabled)
   Active: failed (Result: start-limit-hit) since Wed 2020-11-04 22:37:42 UTC; 12min ago
     Docs: https://github.com/coreos/console-login-helper-messages
  Process: 1754 ExecStart=/usr/libexec/console-login-helper-messages/issuegen (code=exited, status=0/SUCCESS)
 Main PID: 1754 (code=exited, status=0/SUCCESS)
      CPU: 9ms

Nov 04 22:37:42 localhost.localdomain systemd[1]: Starting Generate console-login-helper-messages issue snippet...
Nov 04 22:37:42 localhost.localdomain systemd[1]: Started Generate console-login-helper-messages issue snippet.
Nov 04 22:37:42 localhost.localdomain systemd[1]: console-login-helper-messages-issuegen.service: Consumed 9ms CPU time
Nov 04 22:37:42 localhost.localdomain systemd[1]: console-login-helper-messages-issuegen.service: Start request repeated too quickly.
Nov 04 22:37:42 localhost.localdomain systemd[1]: console-login-helper-messages-issuegen.service: Failed with result 'start-limit-hit'.
Nov 04 22:37:42 localhost.localdomain systemd[1]: Failed to start Generate console-login-helper-messages issue snippet.
Nov 04 22:37:42 localhost.localdomain systemd[1]: console-login-helper-messages-issuegen.service: Start request repeated too quickly.
Nov 04 22:37:42 localhost.localdomain systemd[1]: console-login-helper-messages-issuegen.service: Failed with result 'start-limit-hit'.
Nov 04 22:37:42 localhost.localdomain systemd[1]: Failed to start Generate console-login-helper-messages issue snippet.
```

One simple fix might be a `sleep 1` in the service at the end; we don't need to regenerate more than once a second.  But we should also probably adjust the start limits per the bootupd PR above.

**Comment by dustymabe** [comment]
+1 for adjusting the limits

**Comment by dustymabe** [comment]
I'm seeing this as well. Do we know if there is any condition in particular that is causing it (a high number of interfaces maybe?). We might want to get it in sooner than later.

**Comment by kelvinfan001** [comment]
Yeah the only thing I can think of is a high number of interfaces. I don't know what else would be dropping files in to CLHM's `/run/console-login-helper-messages/issue.d`.

**Comment by dustymabe** [comment]
ok so I'm testing with adding

```
StartLimitIntervalSec=1s
StartLimitBurst=1
```

to `[Unit]` within `console-login-helper-messages-issuegen.service`. It seems to work to limit it to just running once per second, but man does it make the output ugly and possibly causes some inappropriate red flags:


```
[FAILED] Failed to start Generate console-login-helper-messages issue snippet.                                                                                                                                                               
See 'systemctl status console-login-hel…messages-issuegen.service' for details.                                                                                                                                                              
[FAILED] Failed to start Generate console-login-helper-messages issue snippet.                                                                                                                                                               
See 'systemctl status console-login-hel…messages-issuegen.service' for details.                                                                                                                                                              
[FAILED] Failed to start Generate console-login-helper-messages issue snippet.                                                                                                                                                               
See 'systemctl status console-login-hel…messages-issuegen.service' for details.                                                                                                                                                              
[FAILED] Failed to start Generate console-login-helper-messages issue snippet.                                                                                                                                                               
See 'systemctl status console-login-hel…messages-issuegen.service' for details.                                                                                                                                                              
[FAILED] Failed to start Generate console-login-helper-messages issue snippet.                                                                                                                                                               
See 'systemctl status console-login-hel…messages-issuegen.service' for details.
```

but then the service looks good when the system is up: 

```
$ systemctl status console-login-helper-messages-issuegen.service
● console-login-helper-messages-issuegen.service - Generate console-login-helper-messages issue snippet
   Loaded: loaded (/etc/systemd/system/console-login-helper-messages-issuegen.service; enabled; vendor preset: enabled)
   Active: inactive (dead) since Tue 2020-11-17 16:54:38 UTC; 3min 21s ago
     Docs: https://github.com/coreos/console-login-helper-messages
  Process: 1272 ExecStart=/usr/libexec/console-login-helper-messages/issuegen (code=exited, status=0/SUCCESS)
 Main PID: 1272 (code=exited, status=0/SUCCESS)

Nov 17 16:54:38 localhost.localdomain systemd[1]: Starting Generate console-login-helper-messages issue snippet...
Nov 17 16:54:38 localhost.localdomain systemd[1]: Started Generate console-login-helper-messages issue snippet.
```

but there were a whole bunch of failures to get to that point:

```
$ journalctl -b 0 -u console-login-helper-messages-issuegen.service
Nov 17 16:54:34 localhost systemd[1]: Starting Generate console-login-helper-messages issue snippet...
Nov 17 16:54:34 localhost systemd[1]: Started Generate console-login-helper-messages issue snippet.
Nov 17 16:54:34 localhost systemd[1]: console-login-helper-messages-issuegen.service: Start request repeated too quickly.
Nov 17 16:54:34 localhost systemd[1]: console-login-helper-messages-issuegen.service: Failed with result 'start-limit-hit'.
Nov 17 16:54:34 localhost systemd[1]: Failed to start Generate console-login-helper-messages issue snippet.
Nov 17 16:54:34 localhost systemd[1]: console-login-helper-messages-issuegen.service: Start request repeated too quickly.
Nov 17 16:54:34 localhost systemd[1]: console-login-helper-messages-issuegen.service: Failed with result 'start-limit-hit'.
Nov 17 16:54:34 localhost systemd[1]: Failed to start Generate console-login-helper-messages issue snippet.
Nov 17 16:54:35 localhost.localdomain systemd[1]: console-login-helper-messages-issuegen.service: Start request repeated too quickly.
Nov 17 16:54:35 localhost.localdomain systemd[1]: console-login-helper-messages-issuegen.service: Failed with result 'start-limit-hit'.
Nov 17 16:54:35 localhost.localdomain systemd[1]: Failed to start Generate console-login-helper-messages issue snippet.
Nov 17 16:54:35 localhost.localdomain systemd[1]: console-login-helper-messages-issuegen.service: Start request repeated too quickly.
Nov 17 16:54:35 localhost.localdomain systemd[1]: console-login-helper-messages-issuegen.service: Failed with result 'start-limit-hit'.
Nov 17 16:54:35 localhost.localdomain systemd[1]: Failed to start Generate console-login-helper-messages issue snippet.
Nov 17 16:54:35 localhost.localdomain systemd[1]: console-login-helper-messages-issuegen.service: Start request repeated too quickly.
Nov 17 16:54:35 localhost.localdomain systemd[1]: console-login-helper-messages-issuegen.service: Failed with result 'start-limit-hit'.
Nov 17 16:54:35 localhost.localdomain systemd[1]: Failed to start Generate console-login-helper-messages issue snippet.
Nov 17 16:54:35 localhost.localdomain systemd[1]: console-login-helper-messages-issuegen.service: Start request repeated too quickly.
Nov 17 16:54:35 localhost.localdomain systemd[1]: console-login-helper-messages-issuegen.service: Failed with result 'start-limit-hit'.
Nov 17 16:54:35 localhost.localdomain systemd[1]: Failed to start Generate console-login-helper-messages issue snippet.
Nov 17 16:54:35 localhost.localdomain systemd[1]: console-login-helper-messages-issuegen.service: Start request repeated too quickly.
Nov 17 16:54:35 localhost.localdomain systemd[1]: console-login-helper-messages-issuegen.service: Failed with result 'start-limit-hit'.
Nov 17 16:54:35 localhost.localdomain systemd[1]: Failed to start Generate console-login-helper-messages issue snippet.
Nov 17 16:54:38 localhost.localdomain systemd[1]: Starting Generate console-login-helper-messages issue snippet...
Nov 17 16:54:38 localhost.localdomain systemd[1]: Started Generate console-login-helper-messages issue snippet.
```

**Comment by dustymabe** [comment]
> ok so I'm testing with adding
> 
> ```
> StartLimitIntervalSec=1s
> StartLimitBurst=1
> ```
> 

And that only "works" because I've made `StartLimitIntervalSec` much lower compared to the default, so another invocation of the path unit manages to sneak in there and "work". So this isn't the right solution.

**Comment by dustymabe** [comment]
I guess the open question is what we adjust `StartLimitIntervalSec` and `StartLimitBurst` to. Do we

- adjust them so that it gets rate limited sooner
- adjust them so it doesn't get rate limited at all (similar to how it was done for `bootupd`)

**Comment by kelvinfan001** [comment]
What would happen if we don't rate limit it? :) 
That would seem to fix the immediate problem. But I think the real solution would be to get it to generate once per burst (at the end of the burst).

**Comment by dustymabe** [comment]
> What would happen if we don't rate limit it? :)

Yeah I think we need to somewhat disable the default rate limiting here, but we may want to do a high `StartLimitBurst`  over a longer period of time just to prevent it from happening too much if an interface were flapping or something. 

> That would seem to fix the immediate problem. But I think the real solution would be to get it to generate once per burst (at the end of the burst).

Yeah, it would be nice if systemd had a "queue up these requests and collapse them down into a single request, and run once at the end of the interval" feature. :)

**Comment by cgwalters** [comment]
Too much chat not enough code, pushed https://github.com/coreos/console-login-helper-messages/pull/88

**Comment by kelvinfan001** [comment]
Fixed in #89
